Add a way to access request context data in `HttpServerMetricsTagsContributor`

geoand opened the pull request:
- Closes: #47886

geoand committed:
Add a way to access request context data in HttpServerMetricsTagsContributor

Closes: #47886

adutra reviewed (approved):
Thank you for the quick implementation!

gastaldi commented on extensions/micrometer/runtime/src/main/java/io/quarkus/micrometer/runtime/HttpServerMetricsTagsContributor.java:
```diff
@@ -23,5 +23,14 @@ interface Context {
+         * work when {@link HttpServerMetricsTagsContributor#contribute(Context)} is called as the HTTP request has
+         * already gone away.
+         */
+        Object requestContextLocalData(Object key);
```
Would it make sense to add some inferring here to avoid casts? 

```suggestion
        <T> T requestContextLocalData(Object key);
```

Then one could use: `String value = requestLocalData("foo")`

geoand replied:
There is no type to use for inference

gastaldi replied:
I'm talking about the expected return type, as a developer it prevents ugly casts in the code as it can be done in the implementation

geoand replied:
Yeah I know. But that would mean having another argument that captures the type, which I don't think is worth it.

gastaldi replied:
Well, not really, you can simply cast to `(T)` and it should just work ™ :)

```java
        @Override
        public <T> T requestContextLocalData(Object key) {
            return (T) ((HttpServerRequestInternal) request).context().getLocal(key);
        }
```

geoand replied:
Feel free to push a commit with that

gastaldi replied:
Done in [`c9509a5` (#47887)](https://github.com/quarkusio/quarkus/pull/47887/commits/c9509a5ffe7143966c6df86dd1c987fa551b232c)

cescoffier reviewed (approved):
I'm not a big fan of reading duplicated content after requests, but we can find a better way later.

gastaldi committed:
Improve type safety in Micrometer HTTP server metrics

```
- Refactor context local data handling across Micrometer integration to improve type safety with generics.
- Simplify and clean up code by removing unnecessary type casts in `ContextLocalTag`.
```

geoand commented:
@gastaldi actually now that I think of it we don't want that because it ties the type of the key and the type of value.

brunobat commented on integration-tests/micrometer-prometheus/src/main/java/io/quarkus/it/micrometer/prometheus/PathTemplateResource.java:
```diff
@@ -4,10 +4,13 @@
 public class PathTemplateResource {
     @GET
     public String get(@PathParam("value") String value) {
+        ContextLocals.put("context-local", "val-" + value);
```
This is a bad practice. We should not encourage ppl to add tags from path params. Even if they think the nr of different values are limited, they cannot really control it.

geoand replied:
It might be bad practice,  but people shouldn't be limited by accidental technically reasons from doing it.

ThreadLocals should not be used generally,  but that doesn't mean we should prevent their use.

brunobat replied:
Fair, but this should come with a big warning that they will shoot themselves in the foot if they don't consider cardinatility.
See for exampleL [#users > Are you using path parameters? (metrics)](https://quarkusio.zulipchat.com/#narrow/channel/187030-users/topic/Are.20you.20using.20path.20parameters.3F.20.28metrics.29/with/518040545)

I think a warning should be placed in the docs and a in the test.

brunobat replied:
Because in the end, ppl will complain and we will spend time debugging this stuff.

geoand replied:
I would add we should also place a warning in the Javadoc

geoand replied:
If you want to add something in the various places, I'll gladly approve

adutra replied:
For context: my initial use case is not related to path params, it's related to logging the tenant ID associated with the request. And tenant IDs are guaranteed to form a finite set, although they could be like a thousand or so. Are you saying that even this use case is bad practice?

brunobat replied:
Yes. It must not be used like that. It will blow the dimensions exponentially. Remember that for each tenant id you will have different metrics per endpoint, response status and method.
Place that data in the span and later query the APM's span attributes and correlate with exemplars.
